src/types/list.rs

```diff
@@ -493,14 +604,21 @@ impl<'py> Iterator for BoundListIterator<'py> {
 
     #[inline]
     fn next(&mut self) -> Option<Self::Item> {
-        let length = self.length.min(self.list.len());
+        split_borrow!(self, index, length, list);
```

Comment:
Especially if we use my suggestion from above to avoid macros on the critical sections, I think this can just be:


```suggestion
        let Self { index, length, list } = self;
```